File: packages/template-ts/src/value/number.ts
// Number conversion and formatting as defined in docs/spec/data-model.md (VAL-9)
// and docs/spec/functions.md (FUN-20 to FUN-25).

export const MAX_SAFE = 9007199254740991;

// VAL-9: ECMAScript Number::toString.
export function numberToString(value: number): string {
  if (value === 0) return '0';
  return String(value);
}

// Shortest round-trip decimal digits of a finite number: digits without leading zeros,
// and the decimal exponent n such that value = 0.digits × 10^n.
export function shortestDigits(value: number): { negative: boolean; digits: string; exponent: number } {
  const negative = value < 0 || Object.is(value, -0);
  const text = Math.abs(value).toExponential();
  const match = /^(\d)(?:\.(\d+))?e([+-]\d+)$/.exec(text);
  if (!match) throw new Error(`cannot decompose ${text}`);
  let digits = (match[1] as string) + (match[2] ?? '');
  digits = digits.replace(/0+$/, '');
  if (digits === '') digits = '0';
  const exponent = Number(match[3]) + 1;
  return { negative, digits, exponent };
}

// Positional decimal expansion without exponent: integer part and fraction part digit strings.
export function positional(value: number): { negative: boolean; integer: string; fraction: string } {
  if (value === 0) return { negative: false, integer: '0', fraction: '' };
  const { negative, digits, exponent } = shortestDigits(value);
  let integer: string;
  let fraction: string;
  if (exponent <= 0) {
    integer = '0';
    fraction = '0'.repeat(-exponent) + digits;
  } else if (exponent >= digits.length) {
    integer = digits + '0'.repeat(exponent - digits.length);
    fraction = '';
  } else {
    integer = digits.slice(0, exponent);
    fraction = digits.slice(exponent);
  }
  return { negative, integer, fraction };
}

// FUN-22: rounds a positional digit string to the given number of fraction digits,
// half away from zero on the decimal digits.
export function roundDecimal(value: number, decimals: number): { negative: boolean; integer: string; fraction: string } {
  const { negative, integer, fraction } = positional(value);
  if (fraction.length <= decimals) {
    return { negative, integer, fraction: fraction + '0'.repeat(decimals - fraction.length) };
  }
  const roundUp = (fraction.charCodeAt(decimals) as number) >= 0x35;
  let kept = integer + fraction.slice(0, decimals);
  if (roundUp) kept = incrementDigits(kept);
  const splitAt = kept.length - decimals;
  return { negative, integer: kept.slice(0, splitAt) || '0', fraction: kept.slice(splitAt) };
}

function incrementDigits(digits: string): string {
  const chars = digits.split('');
  let index = chars.length - 1;
  while (index >= 0) {
    if (chars[index] === '9') {
      chars[index] = '0';
      index--;
    } else {
      chars[index] = String.fromCharCode((chars[index] as string).charCodeAt(0) + 1);
      return chars.join('');
    }
  }
  return '1' + chars.join('');
}

// FUN-20 to FUN-24.
export function formatNumber(value: number, decimals: number, dec: string, thousands: string): string {
  const rounded = roundDecimal(value, decimals);
  const groups: string[] = [];
  let integer = rounded.integer;
  while (integer.length > 3) {
    groups.unshift(integer.slice(-3));
    integer = integer.slice(0, -3);
  }
  groups.unshift(integer);
  let text = groups.join(thousands);
  if (decimals > 0) text += dec + rounded.fraction;
  const allZero = /^[0.]*$/.test(rounded.integer + rounded.fraction);
  return rounded.negative && !allZero ? '-' + text : text;
}

// FUN-25: the rounded value as a number.
export function roundNumber(value: number, decimals: number): number {
  const rounded = roundDecimal(value, decimals);
  const text = rounded.integer + (rounded.fraction ? '.' + rounded.fraction : '');
  const result = Number(text);
  return rounded.negative && result !== 0 ? -result : result;
}

File: packages/template-go/value/number.go
package value

import (
	"math"
	"strconv"
	"strings"
)

// MaxSafe is the largest integer that a number represents exactly (VAL-2).
const MaxSafe = 9007199254740991

// NumberToString implements VAL-9 (ECMAScript Number::toString).
func NumberToString(x float64) string {
	if x == 0 {
		return "0"
	}
	negative := x < 0
	if negative {
		x = -x
	}
	digits, n := ShortestDigits(x)
	k := len(digits)
	var out string
	switch {
	case k <= n && n <= 21:
		out = digits + strings.Repeat("0", n-k)
	case 0 < n && n <= 21:
		out = digits[:n] + "." + digits[n:]
	case -6 < n && n <= 0:
		out = "0." + strings.Repeat("0", -n) + digits
	default:
		e := n - 1
		sign := "+"
		if e < 0 {
			sign = "-"
			e = -e
		}
		if k == 1 {
			out = digits + "e" + sign + strconv.Itoa(e)
		} else {
			out = digits[:1] + "." + digits[1:] + "e" + sign + strconv.Itoa(e)
		}
	}
	if negative {
		return "-" + out
	}
	return out
}

// ShortestDigits returns the shortest round-trip digits of a positive number and the exponent n
// such that x = 0.digits × 10^n (VAL-9 step 3).
func ShortestDigits(x float64) (digits string, n int) {
	text := strconv.FormatFloat(x, 'e', -1, 64)
	mantissa, exp, _ := strings.Cut(text, "e")
	e, _ := strconv.Atoi(exp)
	digits = strings.Replace(mantissa, ".", "", 1)
	digits = strings.TrimRight(digits, "0")
	if digits == "" {
		digits = "0"
	}
	return digits, e + 1
}

// Positional returns the decimal expansion of a number without exponent.
func Positional(x float64) (negative bool, integer, fraction string) {
	if x == 0 {
		return false, "0", ""
	}
	negative = x < 0 || math.Signbit(x)
	digits, n := ShortestDigits(math.Abs(x))
	switch {
	case n <= 0:
		return negative, "0", strings.Repeat("0", -n) + digits
	case n >= len(digits):
		return negative, digits + strings.Repeat("0", n-len(digits)), ""
	default:
		return negative, digits[:n], digits[n:]
	}
}

// RoundDecimal rounds the positional digits half away from zero (FUN-22).
func RoundDecimal(x float64, decimals int) (negative bool, integer, fraction string) {
	negative, integer, fraction = Positional(x)
	if len(fraction) <= decimals {
		return negative, integer, fraction + strings.Repeat("0", decimals-len(fraction))
	}
	roundUp := fraction[decimals] >= '5'
	kept := integer + fraction[:decimals]
	if roundUp {
		kept = incrementDigits(kept)
	}
	split := len(kept) - decimals
	integer = kept[:split]
	if integer == "" {
		integer = "0"
	}
	return negative, integer, kept[split:]
}

func incrementDigits(digits string) string {
	b := []byte(digits)
	for i := len(b) - 1; i >= 0; i-- {
		if b[i] == '9' {
			b[i] = '0'
			continue
		}
		b[i]++
		return string(b)
	}
	return "1" + string(b)
}

// FormatNumber implements FUN-20 to FUN-24.
func FormatNumber(x float64, decimals int, dec, thousands string) string {
	negative, integer, fraction := RoundDecimal(x, decimals)
	var groups []string
	for len(integer) > 3 {
		groups = append([]string{integer[len(integer)-3:]}, groups...)
		integer = integer[:len(integer)-3]
	}
	groups = append([]string{integer}, groups...)
	text := strings.Join(groups, thousands)
	if decimals > 0 {
		text += dec + fraction
	}
	allZero := strings.Trim(integer+fraction, "0") == ""
	if negative && !allZero {
		return "-" + text
	}
	return text
}

// RoundNumber implements FUN-25.
func RoundNumber(x float64, decimals int) float64 {
	negative, integer, fraction := RoundDecimal(x, decimals)
	text := integer
	if fraction != "" {
		text += "." + fraction
	}
	result, _ := strconv.ParseFloat(text, 64)
	if negative && result != 0 {
		return -result
	}
	return result
}
